lib/services/lock.go

```diff
@@ -0,0 +1,80 @@
+/*
+Copyright 2021 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package services
+
+import (
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/lib/utils"
+
+	"github.com/gravitational/trace"
+)
+
+// UnmarshalLock unmarshals the Lock resource from JSON.
+func UnmarshalLock(bytes []byte, opts ...MarshalOption) (types.Lock, error) {
+	if len(bytes) == 0 {
+		return nil, trace.BadParameter("missing resource data")
+	}
+
+	var lock types.LockV2
+	if err := utils.FastUnmarshal(bytes, &lock); err != nil {
+		return nil, trace.BadParameter(err.Error())
+	}
+	if err := lock.CheckAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	cfg, err := CollectOptions(opts)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	if cfg.ID != 0 {
+		lock.SetResourceID(cfg.ID)
+	}
+	if !cfg.Expires.IsZero() {
+		lock.SetExpiry(cfg.Expires)
+	}
+	return &lock, nil
+}
+
+// MarshalLock marshals the Lock resource to JSON.
+func MarshalLock(lock types.Lock, opts ...MarshalOption) ([]byte, error) {
+	if err := lock.CheckAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	cfg, err := CollectOptions(opts)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	switch lock := lock.(type) {
+	case *types.LockV2:
+		if version := lock.GetVersion(); version != types.V2 {
+			return nil, trace.BadParameter("mismatched lock version %v and type %T", version, lock)
+		}
+		if !cfg.PreserveResourceID {
+			// avoid modifying the original object
+			// to prevent unexpected data races
+			copy := *lock
+			copy.SetResourceID(0)
+			lock = &copy
```

Comment:
This has been just copied over from another marshaller. As the comment mentions the copy is used to prevent data races in connection with the `SetResourceID`, and a shallow copy is apparently sufficient for that purpose.